snuba/web/split.py

```diff
@@ -1,34 +1,153 @@
 import copy
 import math
-from datetime import timedelta
-from typing import Any, List, Optional, Union
+from dataclasses import replace
+from datetime import datetime, timedelta
+from typing import Any, List, Optional, Sequence, Tuple, Union
 
-from snuba import state, util
+from snuba import environment, state, util
 from snuba.clickhouse.query import Query
 from snuba.datasets.plans.split_strategy import QuerySplitStrategy, SplitQueryRunner
+from snuba.query.conditions import (
+    OPERATOR_TO_FUNCTION,
+    BooleanFunctions,
+    binary_condition,
+    in_condition,
+    is_binary_condition,
+)
+from snuba.query.dsl import literals_tuple
+from snuba.query.expressions import Column, Expression, FunctionCall, Literal
 from snuba.request.request_settings import RequestSettings
 from snuba.util import is_condition
+from snuba.utils.metrics.backends.wrapper import MetricsWrapper
 from snuba.web import QueryResult
 
+metrics = MetricsWrapper(environment.metrics, "query.splitter")
+
 # Every time we find zero results for a given step, expand the search window by
 # this factor. Based on the assumption that the initial window is 2 hours, the
 # worst case (there are 0 results in the database) would have us making 4
 # queries before hitting the 90d limit (2+20+200+2000 hours == 92 days).
 STEP_GROWTH = 10
 
 
-def _identify_condition(condition: Any, field: str, operand: str) -> bool:
-    return is_condition(condition) and condition[0] == field and condition[1] == operand
+def _identify_condition(condition: Any, field: str, operator: str) -> bool:
+    return (
+        is_condition(condition) and condition[0] == field and condition[1] == operator
+    )
+
+
+def _get_time_range(
+    query: Query, timestamp_field: str
+) -> Tuple[Optional[Literal], Optional[Literal]]:
+    """
+    Finds the time range for this query. Which means, finds the timestamp condition
+    with the highest datetime literal and the timestamp condition with the smallest
+    and returns the interval in the form of a tuple of Literals. It only looks into
+    first level AND conditions.
+
+    TODO: Consider making this part of the AST manipulation api if there are more use
+    cases. It would require managing a few more corner cases for being part of the api.
+    """
+
+    def get_first_level_conditions(exp: Expression) -> Sequence[Expression]:
+        """
+        Traverse the tree as long as it runs into AND conditions and return all the
+        top level conditions.
+        """
+        if isinstance(exp, FunctionCall) and is_binary_condition(
+            exp, BooleanFunctions.AND
+        ):
+            return [
+                *get_first_level_conditions(exp.parameters[0]),
+                *get_first_level_conditions(exp.parameters[1]),
+            ]
+        else:
+            return [exp]
```

Comment:
I think I don't quite understand what is meant by "top level conditions" or am not following something here. If we have an OR function for example, why do we return `[exp]` (instead of `[]`)? I would have expected it's omitted from the list of top level conditions.

Reply:
top level conditions are a concept of the old representation but we still need to deal with them. Essentially are the list of clauses that are in AND in a Snuba query. Which is the first level of the condition nested array.
Example:
```
[
   ["a", "=", 1],
   ["a", "=", 2]
   [
       ["b", "=", 3],
       ["c", "=", 3], 
   ]
]
```
The top level conditions would be:
```
["a", "=", 1],
   
["a", "=", 2]
   
[
       ["b", "=", 3],
       ["c", "=", 3], 
   ]
```

Reply:
Ok. I assumed you'd omit the 3rd one from that, since you're only concerned with the top level non nested conditions. But I guess your usage skips over them in the following step anyway.

Reply:
The usage indeed skips the third case. I think that decision is use case dependent, the responsibility of `get_first_level_conditions` seems more clear to me if it returns all the top level conditions.